hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/Iterators.scala

```diff
@@ -273,17 +299,26 @@ class RecordMergingFileIterator(logFiles: List[HoodieLogFile],
         val result = recordMerger.merge(curRecord, baseFileReaderAvroSchema, newRecord, logFileReaderAvroSchema, payloadProps)
         toScalaOption(result)
           .map { r =>
-            val schema = HoodieInternalRowUtils.getCachedSchema(r.getRight)
-            val projection = HoodieInternalRowUtils.getCachedUnsafeProjection(schema, structTypeSchema)
-            projection.apply(r.getLeft.getData.asInstanceOf[InternalRow])
+            val data = r.getLeft.getData.asInstanceOf[InternalRow]
+            if (isDeleteData(data)) {
+              null
```

Comment:
No, here is the return type of `optional.map` function.
To better readable, I update the pr in the latest commit to change optional.map to optional.flatmap to avoid Some(null).